### PR TITLE
[react-redux] Added context property to Provider + exported ReactReduxContext.

### PR DESCRIPTION
**Pull request by soerenbf**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
https://github.com/reduxjs/react-redux/blob/master/src/components/Provider.js#L80
https://github.com/reduxjs/react-redux/blob/master/src/index.js#L6
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react-redux/index.d.ts
Comment (mDibyo) [inline types/react-redux/index.d.ts]: Can we use `ReactReduxContext` here?

File: types/react-redux/index.d.ts
Comment (mDibyo) [inline types/react-redux/index.d.ts]: Can we use `ReactReduxContext` here as well?

**Review by mDibyo — CHANGES_REQUESTED** [review]

**Comment by mDibyo** [comment]
@soerenbf I left a review ^. Also, could you update the PR description to better describe the change? (See one of the other PRs for `@types/react-redux` for examples.)

**Comment by mDibyo** [comment]
Partially addresses #31144

**Comment by soerenbf** [comment]
> This PR adds stuff pertaining only to react-redux@6.0.0 (which was released a couple of weeks ago). And react-redux@6.0.0 has a couple of breaking changes (https://github.com/reduxjs/react-redux/releases/tag/v6.0.0).
> 
> So, although this PR by itself does not introduce a breaking change, IMO we should bump the major version of @types/react-redux here. See #25435 for an example of how to do this.
> 
> Also see [#25435 (comment)](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25435#issuecomment-385839174) for why the new major version will need to be 7 (as opposed to 6).

@mDibyo I changes the version comment at the type of the definition file, but is that really enough? It's all i was able to find in the PR you linked that was related to versioning.

**Comment by mDibyo** [comment]
@soerenbf I wanted to play with this by dropping it into a real project, so will take some more time here.

**Comment by mDibyo** [comment]
@AMalininHere Good point! Can you create a separate PR for that?

**Comment by AMalininHere** [comment]
@mDibyo I'll do it after merging that pull request

**Review by mDibyo — CHANGES_REQUESTED** [review]
@soerenbf When releasing v6, all the existing files were copied into the v5 subfolder, and the main folder contained the stuff for v6. Think we need to do a similar thing here.
All 5 files in the main directory should be copied into a v6 subfolder (in the state they were before this PR. ) And then the main directory will contain v7 files.

Also, left a couple of review comments.

**Review by mDibyo — APPROVED** [review]
LGTM.

Could a repo maintainer confirm that we followed the right procedure for releasing a new major version of this package? To summarize:
- We copied all files in `types/react-redux` into `types/react-redux/v6`.
- We made changes to files at `types/react-redux`, and bumped the major version in `types/react-redux/index.d.ts`

**Comment by mDibyo** [comment]
@soerenbf Would you like to add yourself as an author on the package?

**Review by mDibyo — APPROVED** [review]
Just reapproving on latest commit so that typescript-bot knows.

**Comment by DanielRosenwasser** [comment]
Thank you!

File: types/react-redux/index.d.ts
Comment (Kovensky) [inline types/react-redux/index.d.ts]: Why was this a version bump? There's no react-redux 7.0. These are types for react-redux 6.0.

Comment (soerenbf) [reply types/react-redux/index.d.ts]: Says in the comment further down the file as below:

```
// NOTE about the wrong react-redux version in the header comment:
// The actual react-redux version is not 6.0.0, but we had to increase the major version
// to update this type definitions for redux@4.x from redux@3.x.
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25321
```

Comment (DanielRosenwasser) [reply types/react-redux/index.d.ts]: @Kovensky what would you advise is the right course of action? Should we unpublish?

Comment (soerenbf) [reply types/react-redux/index.d.ts]: @Kovensky Simply building on the reasoning from the previous version, but if we find a suitable solution I agree that the better approach would be to follow the library targeted.

**Comment by mDibyo** [comment]
@gfx As it is set up now, if react-redux becomes v7, then the corresponding @types/react-redux has to become v8. Breaking changes must lead to a new major version.

> DefinitelyTyped doesn't follow semver; even if the new types are "breaking" from a TypeScript perspective they're still describing the same major.minor version of the target package.
Though I'm indeed not sure how this interacts when the types the DT types depend on are not themselves in DT...

@Jessidhia types packages definitely need to follow semver. We have had situations in the past on this very package, where people have been burned by upgrading to a new minor version of the package, not realizing that it introduces breaking changes. (see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25321)

Also, react-redux@6 needs react@^16.4, which means people using an earlier version of react will continue using react-redux@5, which means we need to continue maintaining @types/react-redux@6 as a _separate_ code branch. Releasing a new major version is the only mechanism for doing so in the DefinitelyTyped repo (that I know of).

cc'ing @soerenbf @DanielRosenwasser

**Comment by Jessidhia** [comment]
There are even certain things that are not breaking changes for the library (maybe by the library's own policies) but are breaking changes for us. For example, [there are no guarantees about the contents of the `ReactReduxContext`](https://react-redux.js.org/using-react-redux/accessing-store#using-reactreduxcontext-directly).

**Comment by mDibyo** [comment]
@Jessidhia Yeah true. :( But in that case arguably, if people are depending on unstable parts of a library, then they should be ok dealing with unstable types for those parts. So, if `ReactReduxContext` internals get changed without there being a major version bump for react-redux, we can make the same changes to @types/react-redux without a major version bump.

**Pull request merged**
